scala-api/src/test/scala/com/cloudera/livy/scalaapi/ScalaClientTest.scala

```diff
@@ -0,0 +1,248 @@
+      var jarFile = new JarOutputStream(new FileOutputStream(jar))
+      jarFile.putNextEntry(new ZipEntry("test.resource"))
+      jarFile.write("test resource".getBytes("UTF-8"))
+      jarFile.closeEntry()
+      jarFile.close()
+      val future = client.addJar(new URI("file:" + jar.getAbsolutePath()))
+      Thread.sleep(5000)
+      val sFuture = client.submit(
+        context => ScalaClientTest.fileOperation(true, "test.resource", context)
+      )
+      val output = Await.result(sFuture, 10 second)
+      assert(output === "test resource")
+    } finally {
+      if (jar != null) {
+        jar.delete()
+      }
+    }
+  }
+
+  private def configureClient(local: Boolean) = {
+    var conf = ScalaClientTest.createConf(local)
+    var javaClient = new LivyClientBuilder(false).setURI(new URI("rsc:/")).setAll(conf).build()
+    client = javaClient.asScalaClient
+  }
+
+  private def pingJob() = {
+    val future = client.submit { context =>
+      null
+    }
+    val result = Await.result(future, 5 second)
+    assert(result == null)
+  }
+}
+
+class CustomTestFailureException extends RuntimeException {}
+
+object ScalaClientTest {
+
+  def createConf(local: Boolean): Properties = {
+    val conf = new Properties
+    if (local) {
+      conf.put(CLIENT_IN_PROCESS.key, "true")
+      conf.put(SparkLauncher.SPARK_MASTER, "local")
+      conf.put("spark.app.name", "SparkClientSuite Local App")
+    } else {
+      val classpath: String = System.getProperty("java.class.path")
+      conf.put("spark.app.name", "SparkClientSuite Remote App")
+      conf.put(SparkLauncher.DRIVER_MEMORY, "512m")
+      conf.put(SparkLauncher.DRIVER_EXTRA_CLASSPATH, classpath)
+      conf.put(SparkLauncher.EXECUTOR_EXTRA_CLASSPATH, classpath)
+    }
+    conf.put(LIVY_JARS.key, "")
+    conf
+  }
+
+  def fileOperation(isResource: Boolean, fileName: String, context: ScalaJobContext): String = {
+    val arr = Seq(1)
+    val rdd = context.sc.parallelize(arr).map(value => {
+      var inputStream: InputStream = null
+      if (isResource) {
+        val ccl = Thread.currentThread.getContextClassLoader
+        inputStream = ccl.getResourceAsStream(fileName)
+      } else {
+        inputStream = new FileInputStream(SparkFiles.get(fileName))
+      }
+      try {
+        val out = new ByteArrayOutputStream()
+        val buffer = new Array[Byte](1024)
+        var read = inputStream.read(buffer)
+        while (read >= 0) {
+          out.write(buffer, 0, read)
+          read = inputStream.read(buffer)
+        }
+        val bytes = out.toByteArray
+        val result = new String(bytes, 0, bytes.length, UTF_8)
+        result
+      } finally {
+        inputStream.close()
+      }
+    })
+    val action = rdd.collect().head
+    action
+  }
+
+  def helloJob(context: ScalaJobContext): String = "hello"
+
+  def throwExceptionJob(context: ScalaJobContext): CustomTestFailureException = throw new CustomTestFailureException
```

Comment:
nit: return type can be `Unit` here since you're not really returning the exception.